File: coding-test/java-kotlin/src/kr/co/programmers/java/greedy/MakeBigNumbers.java
package kr.co.programmers.java.greedy;

public class MakeBigNumbers {

    public static void main(String[] args) {
        String number = "1924";
        int k = 2;
        System.out.println(new Solution().solution(number, k));
    }

    static class Solution {
        public String solution(String number, int k) {
            StringBuilder sb = new StringBuilder();
            int index = 0;
            int max = 0;
            for(int i=0; i<number.length() - k; i++) {
                max = 0;
                for(int j = index; j<= k+i; j++) {
                    if(max < number.charAt(j)-'0') {
                        max = number.charAt(j)-'0';
                        index = j+1;
                    }
                }
                sb.append(max);
            }
            return sb.toString();
        }
    }
}

File: practice-coding-test/javascript/src/algorithm/greedy/make-big-numbers.js
// @ts-check

/**
 * @param {string} number
 * @param {number} k
 * @returns {string}
 */
function solution(number, k) {
  let answer = ''
  let index = 0
  let max = 0
  for (let i = 0; i < number.length - k; i += 1) {
    max = 0
    for (let j = index; j < k + 1 + i; j += 1) {
      if (+number.charAt(j) > max) {
        max = +number.charAt(j)
        index = j + 1
      }
    }
    answer += max
  }

  return answer
}

function diffSolution(number, k) {
  const answer = []
  let head = 0
  let del = k

  answer.push(number[head++])
  while (head < number.length) {
    if (del && answer[answer.length - 1] < number[head]) {
      answer.pop()
      del--
      continue
    }
    answer.push(number[head++])
  }

  return answer.slice(0, number.length - k).join('')
}

const number = '4177252841'
const k = 4

console.log(solution(number, k))
console.log(diffSolution(number, k))

/**
 * number : 4177252841
 * k : 4
 *
 * answer length: number.length - k
 *
 * 1. 첫번째 자리
 *    - [41772][52841]
 *    - [41772] 중 가장 큰 숫자 : 7
 *    - [41] 제거
 *    - k -= 2
 * 2. 두번째 자리
 *    - [7][725][2841]
 *    - [725] 중 가장 큰 숫자 : 7
 *    - 제거 X
 * 3. 세번째 자리
 *    - [77][252][841]
 *    - [252] 중 가장 큰 숫자 : 5
 *    - [2] 제거
 *    - k -= 1
 * 4. 네번째 자리
 *    - [775][28][41]
 *    - [28] 중 가장 큰 숫자 : 8
 *    - [2] 제거
 *    - k -= 1
 * 5. k === 0
 *    - [775841]
 */
